src/idp/mod.rs

```diff
@@ -130,3 +201,283 @@ impl IdentityProvider {
+        let xml_string = saml_response.to_xml().map_err(Error::XmlGenerationError)?;
+
+        let parser = libxml::parser::Parser::default();
+        let xml_document = parser.parse_string(&xml_string)?;
+
+        let mut context = XmlSecSignatureContext::new()?;
+        if sign_assertions || sign_envelope {
+            let key = XmlSecKey::from_rsa_key_der("server_key", &self.signature_key_der)?;
+            context.insert_key(key);
+            // Updating the document with the correct hash value.
+            context.update_document_id_hash(&xml_document, "ID")?;
+        }
+
+        // Parse the xml into libxml2 format.
+        if sign_assertions {
+            // Checking for key and returning an error.
+            context.sign_assertions(&xml_document)?;
+        }
+
+        // Handling assertion encryption.
+        if need_to_encrypt_assertions {
+            let encryption_key_name = self
+                .encryption_key_name
+                .as_ref()
+                .ok_or(Error::MissingEncryptionKeyName)?;
+            let assertion_encryption_key = self
+                .assertion_encryption_key
+                .as_ref()
+                .ok_or(Error::MissingEncryptionKey)?;
+            let key: Vec<u8> = assertion_encryption_key.rsa()?.public_key_to_pem()?;
+            encrypt_assertions(&xml_document, encryption_key_name.as_str(), &key)?;
+        }
+
+        // Signing document envelope.
+        if sign_envelope {
+            context.sign_document(&xml_document, None)?;
+        }
+
+        Ok(xml_document.to_string())
+    }
+
+    fn update_x509_signatures(
+        &self,
+        saml_response: &mut Response,
+        x509_cert_der: &[u8],
+    ) -> Result<(), Error> {
+        let encoded_cert = crate::crypto::mime_encode_x509_cert(x509_cert_der);
+        saml_response
+            .signature
+            .as_mut()
+            .iter_mut()
+            .for_each(|sig| update_signature(&encoded_cert, sig));
+        saml_response
+            .encrypted_assertions
+            .iter_mut()
+            .for_each(|enc_assertions| {
+                enc_assertions.iter_mut().for_each(|x| {
+                    x.assertion
+                        .signature
+                        .as_mut()
+                        .iter_mut()
+                        .for_each(|sig| update_signature(&encoded_cert, sig));
+                })
+            });
+        saml_response.assertions.iter_mut().for_each(|x| {
+            x.iter_mut().for_each(|x| {
+                x.signature
+                    .as_mut()
+                    .iter_mut()
+                    .for_each(|sig| update_signature(&encoded_cert, sig));
+            })
+        });
+        Ok(())
+    }
+}
+
+fn build_algo_to_key_map() -> std::collections::HashMap<String, usize> {
+    let mut ret = std::collections::HashMap::<String, usize>::new();
+    ret.insert(
+        "http://www.w3.org/2001/04/xmlenc#tripledes-cbc".to_string(),
+        21,
```

Comment:
Should we change to 168 / 8 for consistency?

```suggestion
        168 / 8,
```

Reply:
Fixed.